### PR TITLE
Ports - Shamira

### PR DESCRIPTION
**Pull request by MiraMarshall**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | I was able to apply my logic about conditionals and loops from Ruby. The only difference was the syntax. 
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | No, it seems like there are more Javascript resources online than Ruby. It's nice to know that MDN has a thorough guide to Javascript. 
What was a challenge you were able to overcome on this assignment? | One challenge I overcame was being able to DRY up the drawLetters() function. I didn't want to have to type out every letter multiple times. I was able to find a Javascript method .entries to access the key/value pairs of my object and turn it into an array and loop through that to create the draw letter pool. 
What has been interesting and positive about learning a new programming language? | I like that I am able to troubleshoot on my own more with Javascript. 
What is something to focus on your learnings in JavaScript in the next week? | I would like to learn more about DRYing up my code and being more clever with my logic.

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Great work on this project, Shamira! You have a successful Adagrams solution in JavaScript! \o/ Overall, your logic and submission look great and meet all of the requirements.


I've left a lot of comments just to help out with my thoughts on your code style and syntax. Right now, you have a pretty severe habit of indentation... hehehehe. I've pointed those out (lovingly) along with some ways that the code could be more JavaScript-y.


Overall, well done!

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,106 @@
+import { arrayExpression } from "@babel/types";
```
Comment (tildeee) [inline src/adagrams.js:1]: Hm, it seems like this is never used; maybe you can get rid of this line?

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,106 @@
...
+      // call forEach on entries and add it to letterArray
+      Object.entries(letterPool).forEach(function (entry){
+        const letter = entry[0];
+        const value = entry[1];
+
+        for (let i = 0; i < value; i += 1){
+
+          letterArray.push(letter);
+        }
+      });
```
Comment (tildeee) [inline src/adagrams.js:19]: Nice! Clever!

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,106 @@
...
+      
+      const letterHand = [];
+
+      for (let i = 0; i < 10; i += 1){
+     
+      let randomTen = letterArray[Math.floor(Math.random()*letterArray.length)];
```
Comment (tildeee) [inline src/adagrams.js:25]: Watch your indentation here...!

Comment (tildeee) [inline src/adagrams.js:25]: maybe `randomTen` isn't the best variable name here? Maybe simply `letter` or `randomLetter`?

File: src/adagrams.js [hunk 4]
```diff
@@ -1,8 +1,106 @@
...
+    } 
+    return letterHand;
+    // console.log(letterHand);
+  },
+  
+usesAvailableLetters(input, lettersInHand) {
+  
+   let letters = lettersInHand;
+   const match = true;
+   const noMatch = false;
```
Comment (tildeee) [inline src/adagrams.js:38]: Watch your indentation here! (It's in a "half-indentation" state)

Comment (tildeee) [inline src/adagrams.js:38]: Also, I think in some cases, having a variable name for `match = true` and `noMatch = false` is necessary. Because of the size of this method, I _might_ argue to just use the literal booleans below `return false;` and `return match;`

File: src/adagrams.js [hunk 5]
```diff
@@ -1,8 +1,106 @@
...
+
+  for (let i = 0; i < input.length; i += 1 ){
+    if (letters.includes(input[i])) {
+      letters.splice(input[i], 1);
+    }else {
```
Comment (tildeee) [inline src/adagrams.js:43]: Typical JS code style would have a space on the left of `else`, so `} else {`

File: src/adagrams.js [hunk 6]
```diff
@@ -1,8 +1,106 @@
...
+    return match;
   },
+
+scoreWord(word) {
+  let pointValue = {
+      'A': 1,'E': 1,'I': 1,'O': 1,'U': 1,'L': 1,'N': 1,
+      'R': 1,'S': 1,'T': 1,'D': 2,'G': 2,'B': 3,'C': 3,
+      'M': 3,'P': 3,'F': 4,'H': 4,'V': 4,'W': 4,'Y': 4,
+      'K': 5,'J': 8,'X': 8,'Q': 10,'Z': 10 
+    };
```
Comment (tildeee) [inline src/adagrams.js:56]: `pointValue` can be a `const`!

Comment (tildeee) [inline src/adagrams.js:56]: Turns out, JS defaults keys to strings, so you could actually do `const pointValue = { A: 1, E: 1, I: 1, ...}` etc

File: src/adagrams.js [hunk 7]
```diff
@@ -1,8 +1,106 @@
...
+
+let wordScore = 0;
```
Comment (tildeee) [inline src/adagrams.js:58]: Again, indentation ;)

File: src/adagrams.js [hunk 8]
```diff
@@ -1,8 +1,106 @@
...
+word = word.toUpperCase().split('');
```
Comment (tildeee) [inline src/adagrams.js:59]: instead of re-assigning this modified `word` to the same variable `word`, I _might_ prefer a new variable, like `const letters = word.toUpperCase().split('');`

File: src/adagrams.js [hunk 9]
```diff
@@ -1,8 +1,106 @@
...
+} 
+},
+
+
+highestScoreFrom(words){
+
+let winner = {
+  'word': '',
+  'score': 0,
 };
```
Comment (tildeee) [inline src/adagrams.js]: Watch your indentation in this function!

File: src/adagrams.js [hunk 10]
```diff
@@ -1,8 +1,106 @@
...
+  return wordScore;
+} 
+},
+
+
+highestScoreFrom(words){
+
+let winner = {
+  'word': '',
+  'score': 0,
```
Comment (tildeee) [inline src/adagrams.js:77]: Watch your indentation in this function!

Comment (tildeee) [inline src/adagrams.js:77]: Also, like before, about JS defaulting to strings, it'll work if this looks like this:

```javascript
let winner = {
  word: '',
  score: 0,
}
```

File: src/adagrams.js [hunk 11]
```diff
@@ -1,8 +1,106 @@
...
 };
 
+// Question for Dee. There has to be a more suscinct way of doing all of these conditionals, but I couldn't thing of one. 
+// Would like to talk about it in our one on one.  
+// learned about this for...of statement.
```
Comment (tildeee) [inline src/adagrams.js:82]: The most succinct way to do these conditions would be this:

- For every word in words,
- first check the score. That rules out a lot of cases. If the score of the current word (`word`) is higher than `winner.score`, then you have a clear case: you have a new winner.
- It's only at the time of TIE (aka `this.scoreWord(word) === winner.score`) that you have to start looking at length
- Then you'll need to check within that tie some conditionals about length

However, given the complex rules of the project, I personally always expect it to look a little messy :P

`for...of` is cool by the way!!

File: src/adagrams.js [hunk 12]
```diff
@@ -1,8 +1,106 @@
...
+    winner['score'] = this.scoreWord(word);
+  } else if (this.scoreWord(word) >= winner['score'] && word.length == 10 && winner['word'].length != 10) {
+    winner['word'] = word;
+    winner['score'] = this.scoreWord(word);
+  } else if (this.scoreWord(word) >= winner['score'] && word.length < winner['word'].length && winner['word'].length != 10) {
+    winner['word'] = word;
+    winner['score'] = this.scoreWord(word);
+  } else if (this.scoreWord(word) > winner['score']) {
+    winner['word'] = word;
+    winner['score'] = this.scoreWord(word);
```
Comment (tildeee) [inline src/adagrams.js:96]: it's technically less characters if you use dot-notation ;) so `winner.word` and `winner.score`. I wouldn't fight for this, just a thought
